### PR TITLE
PR - Webpage Updating (Intro, generated code headings, Documentation, icon)

### PR DESCRIPTION
**Pull request by danscime**
This PR updates the current website with a few additions.

- Icon added to webpage.
- "Generated Code" headings only added if there are source code folders for the example.
- Adding an intro paragraph to briefly describe Drasil.
- Adding a brief description for the Haskell Documentation section.

I believe this can close the following issues: #1596, #1597, #1599, and #1633.

We can always reopen or create a new issue if something should be changed.

### CODE REVIEW

File: code/website/site.hs [hunk 1]
```diff
@@ -33,46 +33,101 @@ getExtensionName _ = error "Expected some extension."
...
+  -- here, listFieldWith used with multiple versions of SRS per example
   listFieldWith "srs" (
+    -- field will replace the string parameters in the html with the returned string
+    -- <> acts as a monad concatenator
+
+    -- returns filename from ((filename, TYPE), E _ _ _ _)
     field "filename" (return . fst . srsVar) <>
+    -- returns TYPE from ((_, _), E _ _ _ _)
     field "type" (return . snd . srsVar) <>
+    -- returns name from ((_, _), E name [codeSource] srsVariants description)
```
Comment (Mornix) [inline code/website/site.hs]: I would add here (and really should have when I wrote it, the `name` we define here is the same `name` as the outer scope, but for some reason the templating language doesn't keep the outer variables in scope when we enter a for loop, so we redefine it to make it available there.

File: code/website/site.hs [hunk 2]
```diff
@@ -33,46 +33,101 @@ getExtensionName _ = error "Expected some extension."
...
+
 mkExampleCtx :: FilePath -> FilePath -> Context Example
 mkExampleCtx exampleDir srsDir =
+  -- each function is applied to the item containing an Example of form:
+  -- E Name [CodeSource] SRSVariants Description
+
+  -- here, listFieldWith used with multiple versions of SRS per example
   listFieldWith "srs" (
+    -- field will replace the string parameters in the html with the returned string
+    -- <> acts as a monad concatenator
```
Comment (Mornix) [inline code/website/site.hs]: Not a monad, [semigroup](http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Semigroup.html#v:-60--62-)

File: code/website/site.hs [hunk 3]
```diff
@@ -33,46 +33,101 @@ getExtensionName _ = error "Expected some extension."
...
   val <- f i
+  -- if val is Just and the input string matches the html file paramater,
+  -- then we use the String contained in val
   if k == s && isJust val then
     return $ StringField $ fromJust val
   else
+    -- Otherwise, fail acts to not use the $if$ parameter in the html file
     fail $ "maybeField " ++ s ++ " used when really Nothing. Wrap in `$if(" ++ s ++ ")$` block."
 
+-- List if unempty
```
Comment (Mornix) [inline code/website/site.hs]: Pedantic: Not empty or non-empty

File: code/website/site.hs [hunk 4]
```diff
@@ -33,46 +33,101 @@ getExtensionName _ = error "Expected some extension."
...
+maybeListFieldWith :: String -> Context a -> (Item b -> [c]) -> (Item b -> Compiler [Item a]) -> Context b
+-- takes as input: the string to be replaced, the context to be used as is,
+-- a function to be used on input to return a list to check for emptiness,
+-- and the function to be used regularly on input
+maybeListFieldWith s contxt checkNull f = listFieldWith s contxt
+  -- If the list is empty, fail is used so as not to create a specific section
+  (\x -> if null (checkNull x) then fail "No code files for example" else f x)
```
Comment (Mornix) [inline code/website/site.hs]: Error message should be more generic.

File: code/website/index.html [hunk 5]
```diff
@@ -3,11 +3,18 @@ <h2>Drasil - Generate All the Things!</h2>
...
   This webpage is designed to contain the most up to date
-  <a href="#caseStudies">case study artifacts</a>
+  <a href="#caseStudies">case study artifacts</a>, <a href="#haskDoc">documentation</a>,
   and <a href="#packages">package dependency graphs</a> from the Drasil repository.
   The case study artifacts include the Software Requirements Specification (SRS) for the case study,
   which specifies what the program sets out to achieve.
+  The Haskell Documentation section contains the current documentation for the Drasil framework.
   The package dependency graphs shows the hierarchy of modules within each package.
   The footer of this page contains the continuous integration build of the project,
   as well as the commit number that the build and artifacts are based off of.
```
Comment (Mornix) [inline code/website/index.html]: The changes for documentation prompted this comment, but we should have done it before. Any section which is `if`d below (like documentation), should have the same guards in the section, so we don't generate a link to a non-existent documentation section when not present, and similarly don't describe it.

File: code/website/index.html [hunk 6]
```diff
@@ -27,17 +34,28 @@ <h3 id="caseStudies">Case Studies</h3>
...
   </li>$sep-$
 $endfor$
 </ul>
+
 $if(docsUrl)-$
-<h4><a href="$docsUrl$">Haskell Documentation</a></h4>
+  <h3 id="haskDoc">Haskell Documentation</h3>
+  <p>
+    The current documentation for the Drasil framework can be found at
+    <a href="$docsUrl$">this link</a>.
```
Comment (Mornix) [inline code/website/index.html]: Suggestion: I'm not a fan of "this link" (even though I'm occasionally guilty of doing it); maybe something like "The (link)Haddock Documentation(/link) for the Drasil framework." @JacquesCarette or @smiths may have other suggestion or preferences though.

File: code/scripts/deploy_stage.sh [hunk 7]
```diff
@@ -71,11 +71,17 @@ copy_examples() {
 		if [ -d "$example/"src ]; then
 			# We don't expose code in deploy. It's more conveneient to link to GitHub's directory
 			# We place a stub file which Hakyll will replace.
-			echo $(cd "$CUR_DIR" && "$MAKE" deploy_code_path | grep "$example_name" | cut -d"$DEPLOY_CODE_PATH_KV_SEP" -f 2-) > "$EXAMPLE_DEST$example_name/src"
+      REL_PATH=$(cd "$CUR_DIR" && "$MAKE" deploy_code_path | grep "$example_name" | cut -d"$DEPLOY_CODE_PATH_KV_SEP" -f 2-)
+      ls -d "$(git rev-parse --show-toplevel)/$REL_PATH"*/ | rev | cut -d/ -f2 | rev | tr '\n' '\0' | xargs -0 printf "$REL_PATH%s\n" > "$EXAMPLE_DEST$example_name/src"       #echo $(cd "$CUR_DIR" && "$MAKE" deploy_code_path | grep "$example_name" | cut -d"$DEPLOY_CODE_PATH_KV_SEP" -f 2-) > "$EXAMPLE_DEST$example_name/src"
```
Comment (Mornix) [inline code/scripts/deploy_stage.sh]: Should remove the commented portion from the end of this line.

File: code/scripts/deploy_stage.sh [hunk 8]
```diff
@@ -71,11 +71,17 @@ copy_examples() {
...
 		fi
 	done
 }
 
+copy_images() {
+  mkdir -p "$CUR_DIR"deploy/Images
+  cp -r "$CUR_DIR"website/Images/ "$CUR_DIR"deploy/Images
```
Comment (Mornix) [inline code/scripts/deploy_stage.sh]: Pedantic: I personally find capital letters in URL paths funny looking.

File: code/website/index.html (hunk 6 above)
Comment (smiths) [reply code/website/index.html]: I agree with @Mornix.  "this link" doesn't tell you what it is if you just scan the links.  I like the Haddock Documentation suggestion.

File: code/website/index.html (hunk 5 above)
Comment (danscime) [reply code/website/index.html]: So far, Documentation is the only `if` that is also used in the intro. 

Case study artifacts and dependency graphs use `for` so I can not make this change for them (unless I rebuild them using the new `maybeListFieldWith`.

File: code/website/index.html (hunk 6 above)
Comment (danscime) [reply code/website/index.html]: Currently I have:
`The current <a href="$docsUrl$">Haddock documentation</a> for the Drasil framework.`

File: code/website/index.html (hunk 5 above)
Comment (Mornix) [reply code/website/index.html]: Adding a conditional for the preamble about the docs should be sufficient.

File: code/website/index.html (hunk 6 above)
Comment (JacquesCarette) [reply code/website/index.html]: Good - I completely agree that "this link" style is bad.

File: code/website/index.html [hunk 9]
```diff
@@ -3,11 +3,19 @@ <h2>Drasil - Generate All the Things!</h2>
...
+  knowledge should be used in which artifacts, how to transform them, and more.
+  <br><br>
   This webpage is designed to contain the most up to date
-  <a href="#caseStudies">case study artifacts</a>
+  <a href="#caseStudies">case study artifacts</a>,
+  $if(docsUrl)$ <a href="#haskDoc">Haddock documentation</a>, $endif$
   and <a href="#packages">package dependency graphs</a> from the Drasil repository.
   The case study artifacts include the Software Requirements Specification (SRS) for the case study,
   which specifies what the program sets out to achieve.
+  The Haskell Documentation section contains the current documentation for the Drasil framework.
```
Comment (JacquesCarette) [inline code/website/index.html]: Haskell -> Haddock.  Lack of traceability / coherence caused this inconsistency to appear!

Comment (danscime) [reply code/website/index.html]: Fixed in 4528e776df7bc7133e8046640ebca8280681222e

**Pull request merged**